Funcky/Extensions/EnumerableExtensions/AverageOrNone.cs

```diff
@@ -0,0 +1,320 @@
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of values that are used to calculate an average.</param>
+        /// <param name="selector">A transform function to apply to each element.</param>
+        /// <returns>The average of the sequence of values or None.</returns>
+        [Pure]
+        public static Option<decimal> AverageOrNone<TSource>(this IEnumerable<TSource> source, Func<TSource, decimal> selector)
+            => source.Aggregate(AverageCalculatorDecimal.Empty, (calculator, element) => calculator.Add(selector(element))).Average;
+
+        /// <summary>
+        /// Computes the average of a sequence of optional Decimal values that are obtained by invoking a transform function on each element of the input sequence.
+        /// If the sequence only consists of none or is empty it returns None.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the elements of source.</typeparam>
+        /// <param name="source">A sequence of values that are used to calculate an average.</param>
+        /// <param name="selector">A transform function to apply to each element.</param>
+        /// <returns>The average of the sequence of values or None.</returns>
+        [Pure]
+        public static Option<decimal> AverageOrNone<TSource>(this IEnumerable<TSource> source, Func<TSource, Option<decimal>> selector)
+            => source.Aggregate(AverageCalculatorDecimal.Empty, (calculator, element) => calculator.Add(selector(element))).Average;
+
+        private class AverageCalculatorDouble
+        {
+            public static readonly AverageCalculatorDouble Empty = new AverageCalculatorDouble();
+
+            private readonly int _count;
+            private readonly Option<double> _sum;
+
+            public AverageCalculatorDouble(int count = default, Option<double> sum = default)
+            {
+                _count = count;
+                _sum = sum;
+            }
+
+            public Option<double> Average => _sum.Select(sum => sum / _count);
+
+            public AverageCalculatorDouble Add(int term)
+                => new AverageCalculatorDouble(_count + 1, Some(_sum.Match(none: term, some: sum => sum + term)));
+
+            public AverageCalculatorDouble Add(Option<int> term)
+                => term.Match(this, Add);
+
+            public AverageCalculatorDouble Add(long term)
+                => new AverageCalculatorDouble(_count + 1, Some(_sum.Match(none: term, some: sum => sum + term)));
+
+            public AverageCalculatorDouble Add(Option<long> term)
+                => term.Match(this, Add);
+
+            public AverageCalculatorDouble Add(double term)
+                => new AverageCalculatorDouble(_count + 1, Some(_sum.Match(none: term, some: sum => sum + term)));
+
+            public AverageCalculatorDouble Add(Option<double> term)
+                => term.Match(this, Add);
```

Comment:
Add argument labels (applies to all `Calculator`s)

Reply:
Adressed in 9bebfce5